### PR TITLE
Remove version dependence from codegen

### PR DESCRIPTION
**Pull request by fbs**
sorry for the noisy PR. All the magic happens in `codegen/llvm/common.h`

----

The codegen tests are currently quite tedious to update as the "source"
and "result" are defined in the same file which makes it hard to
automatically update.

This commit changes that by:

- Moving the "result" to its own file, `TEST_NAME.ll`
- Introducing a rewriter that tries to automatically fix the LLVM
  version differences
- As rewriting doesn't always work it supports loading results from
  `TEST_NAME_LLVM-X.ll` these will have to be maintained manually but
  at the moment there are only a few (only LLVM 5)

### CODE REVIEW

**Comment by fbs** [comment]
This  hopefully makes  adding llvm-9 codegen tests a  lot easier too. I will give that a  try  once (if) this gets merged

File: tests/codegen/common.h [hunk 1]
```diff
@@ -18,17 +18,151 @@ namespace bpftrace {
...
+  // clang-format on
+  auto parts = split_string(line, ' ');
+  std::stringstream buf;
+  std::string alignment = "";
+  for (std::vector<std::string>::iterator it = parts.begin(); it != parts.end();
+       it++)
+  {
+    if (*it == "align")
+    {
+      if (alignment == "")
```
Comment (fbs) [inline tests/codegen/common.h]: This is a bit of a hack as we have a few tests with a funky address space, so only the value after the first align is valid:

```
call void @llvm.memcpy.p0i8.p64i8.i64(i8* nonnull align 1 %1, i8 addrspace(64)* align 536870912 null, i64 32, i1 false)
```

File: tests/codegen/common.h [hunk 2]
```diff
@@ -18,17 +18,151 @@ namespace bpftrace {
...
+{
+#if LLVM_VERSION_MAJOR < 7
+  // clang-format off
+  //FROM: call void @llvm.memcpy.p0i8.p0i8.i64(i8* nonnull %4, i8* nonnull align 1 i32 1, %lookup_elem, i64 32, i1 false)
+  //TO:  call void @llvm.memcpy.p0i8.p0i8.i64(i8* nonnull %4, i8* nonnull %lookup_elem, i64 32, i32 1, i1 false)
+  // clang-format on
+  auto parts = split_string(line, ' ');
+  std::stringstream buf;
+  std::string alignment = "";
+  for (std::vector<std::string>::iterator it = parts.begin(); it != parts.end();
```
Comment (danobi) [inline tests/codegen/common.h]:
```suggestion
  for (auto it = parts.begin(); it != parts.end();
```

File: tests/codegen/common.h [hunk 3]
```diff
@@ -18,17 +18,151 @@ namespace bpftrace {
...
+static std::string rewrite_memcpy_call(const std::string &line)
+{
+#if LLVM_VERSION_MAJOR < 7
+  // clang-format off
+  //FROM: call void @llvm.memcpy.p0i8.p0i8.i64(i8* nonnull %4, i8* nonnull align 1 i32 1, %lookup_elem, i64 32, i1 false)
+  //TO:  call void @llvm.memcpy.p0i8.p0i8.i64(i8* nonnull %4, i8* nonnull %lookup_elem, i64 32, i32 1, i1 false)
+  // clang-format on
+  auto parts = split_string(line, ' ');
+  std::stringstream buf;
+  std::string alignment = "";
```
Comment (danobi) [inline tests/codegen/common.h]: Default initializes to empty
```suggestion
  std::string alignment;
```

File: tests/codegen/common.h (hunk 1 above)
Comment (danobi) [reply tests/codegen/common.h]: Maybe add a comment mentioning it

Comment (danobi) [inline tests/codegen/common.h]: Nit
```suggestion
      if (alignment.empty())
```

File: tests/codegen/common.h [hunk 4]
```diff
@@ -18,17 +18,151 @@ namespace bpftrace {
...
+  parts.insert(parts.end() - 2, std::string("i32 ") + alignment + ",");
+
+  buf << parts.at(0);
+  for (unsigned int i = 1; i < parts.size(); i++)
+  {
+    buf << " " << parts.at(i);
+  }
+  return buf.str();
+#else
+  return std::string(line);
```
Comment (danobi) [inline tests/codegen/common.h]:
```suggestion
  return line;
```

File: tests/codegen/common.h [hunk 5]
```diff
@@ -18,17 +18,151 @@ namespace bpftrace {
...
+{
+#if LLVM_VERSION_MAJOR < 7
+  if (line.find("addrspace") != std::string::npos)
+    return std::string(
+        "declare void @llvm.memcpy.p0i8.p64i8.i64(i8* nocapture writeonly, i8 "
+        "addrspace(64)* nocapture readonly, i64, i32, i1) #1");
+  return std::string("declare void @llvm.memcpy.p0i8.p0i8.i64(i8* nocapture "
+                     "writeonly, i8* nocapture readonly, i64, i32, i1) #1");
+#else
+  return std::string(line);
```
Comment (danobi) [inline tests/codegen/common.h]:
```suggestion
  return line;
```

File: tests/codegen/common.h [hunk 6]
```diff
@@ -18,17 +18,151 @@ namespace bpftrace {
...
+#endif
+}
+
+static std::string rewrite(const std::string &ir)
+{
+  std::stringstream buf;
+  for (auto &line : split_string(ir, '\n'))
```
Comment (danobi) [inline tests/codegen/common.h]:
```suggestion
  for (const auto &line : split_string(ir, '\n'))
```

**Comment by fbs** [comment]
@danobi  thanks for the review. I've updated it to have the test runner update the codegen when `BPFTRACE_UPDATE_TESTS` is set to `1` in the env.  Still not the best user experience when updating tests but it should be a lot better than is it now

**Comment by danobi** [comment]
Oh, also please add documentation. Maybe in the README.md in tests/

**Comment by fbs** [comment]
@danobi / @ajor  I've added docs and added  LLVM9 support (which turned out quite trivial). Can you take a look at last 2 commits? 

https://github.com/iovisor/bpftrace/pull/1108/commits/6036494814b6a6bbd3fc3eb4474f0f763f87c33f

**Comment by fbs** [comment]
@mmisono  any objection to me merging this or would you rather wait until #1104 is merged as it impacts  your work? It's blocking #1105 and testing on fedora >=31 systems.

**Comment by mmisono** [comment]
@fbs
I'm glad if #1104 is merged first, but not sure how long it takes, so please go ahead. This and #1105 are important changes and I also want them to be merged.

By the way, I think it is useful if bpftrace has an option to save LLVM IR to a file like [this --elf option](https://github.com/iovisor/bpftrace/pull/871/commits/c97ce70aada26e63b1a49fe2c36c2475b83).

**Comment by fbs** [comment]
> @fbs
> I'm glad if #1104 is merged first, but not sure how long it takes, so please go ahead. This and #1105 are important changes and I also want them to be merged.

I'll hold it off for the weekend then :)

> By the way, I think it is useful if bpftrace has an option to save LLVM IR to a file like [this --elf option](https://github.com/iovisor/bpftrace/pull/871/commits/c97ce70aada26e63b1a49fe2c36c2475b83).

I had something like this in #1066 but there is some discussion still about how to name the flags.  I got a WIP somewhere that adds `--emit-<ir,opt-ir,elf> [filename]` somewhere, i'll try and clean that up and open it as a rfc for discussion :)

**Pull request merged**